crates/matrix-sdk/tests/integration/client.rs

```diff
@@ -377,3 +377,56 @@ async fn room_update_channel() {
     assert_eq!(updates.unread_notifications.highlight_count, 0);
     assert_eq!(updates.unread_notifications.notification_count, 11);
 }
+
+// Check that the `Room::is_encrypted()` is properly deduplicated, meaning we
+// only make a single request to the server, and that multiple calls do return
+// the same result.
+#[cfg(all(feature = "e2e-encryption", not(target_arch = "wasm32")))]
+#[async_test]
+async fn request_encryption_event_before_sending() {
+    let room_id = &test_json::DEFAULT_SYNC_ROOM_ID;
+    let (client, server) = logged_in_client().await;
+
+    mock_sync(&server, &*test_json::SYNC, None).await;
+    client
+        .sync_once(SyncSettings::default())
+        .await
+        .expect("We should be able to performa an initial sync");
+
+    let room = client.get_room(room_id).expect("We should know about our default room");
+
+    Mock::given(method("GET"))
+        .and(path_regex(r"^/_matrix/client/r0/rooms/.*/state/m.room.encryption/"))
+        .and(header("authorization", "Bearer 1234"))
+        .respond_with(
+            ResponseTemplate::new(200)
+                .set_body_json(json!({
+                    "algorithm": "m.megolm.v1.aes-sha2",
+                    "rotation_period_ms": 604800000,
+                    "rotation_period_msgs": 100
+                }))
+                // Introduce a delay so the first `is_encrypted()` doesn't finish before we make
+                // the second call.
+                .set_delay(Duration::from_millis(50)),
+        )
+        .mount(&server)
+        .await;
+
+    let first_handle = tokio::spawn({
+        let room = room.to_owned();
+        async move { room.to_owned().is_encrypted().await }
+    });
+
+    let second_handle = tokio::spawn(async move { room.is_encrypted().await });
+
+    let first_encrypted =
+        first_handle.await.unwrap().expect("We should be able to test if the room is encrypted.");
+    let second_encrypted =
+        second_handle.await.unwrap().expect("We should be able to test if the room is encrypted.");
```

Comment:
For more chaos, can you use `join_all`?

Reply:
Not sure if this is nicer, but added in https://github.com/matrix-org/matrix-rust-sdk/pull/2405/commits/de97d034d2a9a3b8746b952f17919fe576cd5894.

Reply:
Ah right, I realize I misunderstood as I wanted to make sure the requests ran concurrently; but this wasn't needed because of the `spawn()`. So feel free to revert if needs be :+1: